src/controllers/helpers/projects/projectHelpers.js

```diff
@@ -0,0 +1,169 @@
+import {
+  showTodoBody,
+  extractProject
+} from '../todos/todoBody';
+import projectsController from "../../projectsController";
+
+import {
+  generateAddTodoBtn
+} from "../todos/todoHelpers";
+
+export const showProjectList = () => {
+  // show project list only if there are projects in the projectsArray
+  let projectsArray = JSON.parse(localStorage.getItem("projectsArray"));
+  if (projectsArray.length > 0) {
+    // Display should show all projects in li's
+    projectsArray.forEach(project => {
+      appendNewProject(project);
+    });
+  } else {
+    // The display should show "No Projects Yet"
+    let li = document.createElement("li");
+    li.setAttribute("class", "list-group-item emptyMessage");
+    li.innerText = "No projects yet, create one";
+    document.getElementsByTagName("ul")[0].appendChild(li);
+  }
+}
+
+// Function to append created project into the DOM
+export const appendNewProject = project => {
+  // Remove any empty project messages
+  removeEmptyProjMessage();
+  // Generate project li
+  generateProjectLi(project);
+};
+
+// Function to remove any empty project message
+const removeEmptyProjMessage = () => {
+  let emptyLi = document.getElementsByClassName("emptyMessage")[0];
+  if (emptyLi !== undefined) {
+    document.getElementsByTagName("ul")[0].removeChild(emptyLi);
+  }
+};
+
+//  Generate project li's
+const generateProjectLi = project => {
+  let li = document.createElement("li"),
+    span = document.createElement("span"),
+    deleteBtn = generateBtn("delete", project),
+    updateBtn = generateBtn("update", project);
+
+  // Give span an id based on project's id and an innerText equal to project's name
+  span.setAttribute("id", `projectSpan-${project.id}`);
+  span.innerText = project.name;
+
+  // Give each li a class of list-group-item and id based on project's id
+  li.setAttribute("class", "list-group-item");
+  li.setAttribute("id", `projectLi-${project.id}`);
+
+  // Add click event listener to each li
+  li.addEventListener('click', () => {
+    // Override the existing project with updated project
+    project = JSON.parse(localStorage.projectsArray).find(x => x.id === project.id);
```

Comment:
Same as previous comment. This helper should not access the DB implementation, rather an abstraction of it.

Reply:
solved